src/FakeItEasy/Configuration/AfterMustHaveHappenedConfiguration.cs

```diff
@@ -17,11 +17,38 @@ public AfterMustHaveHappenedConfiguration(FakeManager fakeManager, ICallMatcher
             this.repeatConstraint = repeatConstraint;
         }
 
-        public void InOrder(ISequentialCallContext context)
+        /// <summary>
+        /// Checks that the asserted call happened in order relative to others in the assertion chain.
+        /// </summary>
+        /// <param name="nextAssertion">An assertion describing the next call that should occur.</param>
+        /// <returns>An object that can be used to assert that a following call was made in the expected order.</returns>
+        /// <exception cref="ExpectationException">The call was not made in the expected order.</exception>
+        public IOrderableCallAssertion Then(IAfterMustHaveHappenedConfiguration nextAssertion)
         {
-            Guard.AgainstNull(context, "context");
+            var context = ServiceLocator.Current.Resolve<SequentialCallContext>();
+            this.CheckCallHappenedInOrder(context);
+            return new SubsequentOrderedCallAssertion(context).Then(nextAssertion);
+        }
 
+        private void CheckCallHappenedInOrder(SequentialCallContext context)
+        {
             context.CheckNextCall(this.fakeManager, this.matcher.Matches, this.callDescription, this.repeatConstraint);
         }
+
+        private class SubsequentOrderedCallAssertion : IOrderableCallAssertion
+        {
+            private readonly SequentialCallContext context;
+
+            public SubsequentOrderedCallAssertion(SequentialCallContext context)
+            {
+                this.context = context;
+            }
+
+            public IOrderableCallAssertion Then(IAfterMustHaveHappenedConfiguration nextCall)
+            {
+                ((AfterMustHaveHappenedConfiguration)nextCall).CheckCallHappenedInOrder(this.context);

```

Comment:
About the downcast, exposing `CheckCallHappenedInOrder` and asking people not to use it is my least favourite of the options. It makes for a dirty API (even though we would be going to lengths to keep the method from being _easily_ accessible). If there's going to be an irregularity, and notes like "don't use this, I mean it", I'd rather they were confined to the internals.

We have liked to return interfaces, but I don't see a compelling reason to when we have a single implementation and there's an advantage to not returning an interface. We could claim it's for extensibility, but I'm not sure there's a huge advantage there. While we _may_ want to add another method to the class, that can be done with a class as well as an interface (and is technically breaking for an interface, but not for a class, so long as the method isn't abstract), I don't think there's much we can do to the ordered assertions feature without requiring a rework of the API (something that would allow us to know when the last ordered assert is made). 

As noted before, we could return the concrete class. I'm not sure what's so terrible about this. If we decide to add another method, we could. If we decide to subclass and split the implementation, we could. If making a method virtual after the fact is considered a breaking change (I don't think it should be), we could make `Then` virtual now.

A variation on "return an interface" or "return a concrete class" is to "return an abstract class", which is nearly the same interface, from the outside. Then `CheckCallHappenedInOrder` can stay on the class, internal. It doesn't bother the users, and we don't have to have a scary cast. I'd had this implemented initially, but didn't see a technical advantage over either returning an interface or a concrete class. In this case, the abstract class would take the place of `IAfterMustHaveHappenedConfiguration`.

I prefer returning some sort of class to returning an interface.

However, to make returning an interface and casting, more palatable, if that's a front-runner for you guys, we could do the following:

1. comments explaining why we're casting and not to add additional production-ready implementations of `IAfterMustHaveHappenedConfiguration`
2. to augment the comments, we could even add a unit test to ensure that nobody adds an implementation unawares